### ISSUE TITLE
Enhancing Router Support for Shared Pages in Monorepo Setups

### ISSUE THREAD

**Issue by n-can-k** [analogjs/analog#904]
### Which scope/s are relevant/related to the feature request?

router

### Information

I'm working on a monorepo (NX) that includes multiple applications, and I've encountered a challenge with routing for pages shared across these applications. Currently, the router doesn't seem to support monitoring pages outside of the app folder, which is crucial for my setup since the apps are essentially just shells, and all modules and pages are imported from the corresponding folder under the `libs`.

To address this, I wanted to implement a custom router provider.
This required me to access to some internal functions like `createRoutes` and `updateMetaTagsOnRouteChange`, so I exported them in my fork. And it now works on my apps.
( I submitted a merge request to expose those internals: https://github.com/analogjs/analog/pull/903 )

Alternatively, I considered extending `provideFileRouter` with a new `files` parameter, but I think this isn't the right approach since it won't be backwards compatible.

Do you think the custom router provider path is the way to go, or is there another solution?

Thanks a ton for considering my two cents.

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by brandonroberts** [comment]
I think exposing `createRoutes` is fine, but I'd rather expose the `ENVIROMENT_INITIALIZER` that `updateMetaTagsOnRouteChange` uses as a function than that function directly. Something like `provideRouteMetaTagsListener()` or something similar.

**Comment by n-can-k** [comment]
i think adding new provider function like this and exposing the `createRoutes` function to create new routes is also an option.

```ts
export function provideFileRouterWithCustomRoutes(
  customRoutes : Route[],
  ...features: RouterFeatures[]
): EnvironmentProviders {
  return makeEnvironmentProviders([
    // TODO: remove type casting after Angular >=15.1.1 upgrade
    // https://github.com/angular/angular/pull/48720
    (
      provideRouter([ ...customRoutes, ...routes ], ...features) as unknown as {
        ɵproviders: Provider[];
      }
    ).ɵproviders,
    {
      provide: ENVIRONMENT_INITIALIZER,
      multi: true,
      useValue: () => updateMetaTagsOnRouteChange(),
    },
  ]);
}
```


this way exposed internals would be minimal

**Comment by brandonroberts** [comment]
You can do that already without anything custom using `provideRoutes`. The `ROUTES` token is uses is a multi-token.

```ts
import {
  provideHttpClient,
  withFetch,
  withInterceptors,
} from '@angular/common/http';
import { ApplicationConfig } from '@angular/core';
import { provideClientHydration } from '@angular/platform-browser';
import { provideFileRouter } from '@analogjs/router';
import { ROUTES, withNavigationErrorHandler } from '@angular/router';
import { cookieInterceptor } from './interceptors/cookies.interceptor';

export const appConfig: ApplicationConfig = {
  providers: [
    { provide: ROUTES, useValue: customRoutes }, // registers array of routes
    provideFileRouter(withNavigationErrorHandler(console.error)), // registers array of routes
    provideHttpClient(withFetch(), withInterceptors([cookieInterceptor])),
    provideClientHydration(),
  ],
};
```

The routes get merged into a single array on that injector.

**Comment by n-can-k** [comment]
Thanks! I didn't know the ROUTES token functioned as a multi-token.

Looks like we just need to expose the `createRoutes` function for now.

Also, we're gonna need to expose a couple of types; `Files` (from './lib/routes') and `RouteExport` (from './lib/models'). They're needed to pass route objects between shared libraries and the app. Is it OK to go ahead with exposing these types?

With this setup, we can start using shared page routers that can be injected from the shared libs, which i think is pretty cool.

**Comment by n-can-k** [comment]
And...

There's a bit of a twist with the `API` (nitro) and `tRPC` routes.
I managed to get all of them working, but the method I used is kinda hacky.

I'm not sure if there's a more graceful way to handle this, but here's how I did it:

1. Pages Routes
1.1. Each lib exports their page routes in `modules/MY-MODULE/src/index.ts` ( this is where `Files` type is used )
1.2. Page routes are imported for each lib and merged in `apps/MY-APP/src/import-routes-pages.ts`
1.3. Import merged routes and create `createRoutes` and provide router in `apps/MY-APP/src/app/app.config.ts`

apps/MY-APP/import-routes-pages.ts:
```ts
import { mergePageRoutes } from '@libs/mod-man';

import { mPages as mPagesA } from '@modules/mod-a';
import { mPages as mPagesB } from '@modules/mod-b';
import { mPages as mPagesNotes } from '@modules/notes';

export const PAGE_ROUTES = mergePageRoutes(mPagesA, mPagesB, mPagesNotes);
```

modules/MY-MODULE/src/index.ts:
```ts
import { RouteExport } from '@analogjs/router';
import { prepareRouteFileObject, is_nitro } from '@libs/mod-man';

const pagesRaw = is_nitro
  ? {}
  : import.meta.glob<RouteExport>(['./pages/**/*.page.ts']);

export const mPages = prepareRouteFileObject(pagesRaw);
```
here `is_nitro` is required because when SSR is running (when in nitro's env) `import.meta.glob` (which is vite's tool) isn't available. `prepareRouteFileObject` is just adding a '/' to the object keys to make them processable by `createRoutes`

apps/MY-APP/src/app/app.config.ts
```ts
...
import { withNavigationErrorHandler, withComponentInputBinding, ROUTES } from '@angular/router';
import { provideFileRouter, createRoutes } from '@analogjs/router';

import { PAGE_ROUTES } from '../../import-routes-pages';
const customRoutes = createRoutes( PAGE_ROUTES );

export const appConfig: ApplicationConfig = {
  providers: [
    { provide: ROUTES, useValue: customRoutes, multi: true }, // registers array of routes
    provideFileRouter(),
...
  ],
};
...
```

2. API (nitro) Routes
2.1. The only way i could find to inject new API routes to Nitro is by overriding `scanDirs` configuration in `apps/MY-APP/vite.config.ts`.
2.2. Since we're in Vite's env now, I couldn't make aliases from tsconfig work. Imported files can't access aliases either. So, to collect required routes from shared libs, i had to provide paths myself. 

apps/MY-APP/import-routes-api.ts:
```ts
import { normalizePath } from 'vite';
import * as path from 'path';

import * as TSConfig from '../../tsconfig.base.json';
const paths = TSConfig.compilerOptions.paths;

export const aliases: { [key: string]: string } = {};
for (const [name, resolution] of Object.entries(paths)) {
  if (!name.startsWith('@modules/') && !name.startsWith('@epmr/auth')) continue;
  aliases[name] = normalizePath(path.resolve(process.cwd(), resolution[0]));
}

export const API_ROUTES = [
  aliases['@modules/mod-a/server'],
  aliases['@modules/mod-b/server'],
  aliases['@modules/notes/server'],
];
```
`@modules/MY-MODULE/server` is a secondary alias for server exports.

apps/app-a/vite.config.ts:
```ts
...
import { defineConfig, normalizePath, splitVendorChunkPlugin } from 'vite';
import { typescriptPaths } from 'rollup-plugin-typescript-paths';

import { API_ROUTES } from './import-routes-api';

export default defineConfig(({ mode }) => {
  return {
...
    plugins: [
      analog({
        nitro: {
          scanDirs: [...API_ROUTES, normalizePath(`${__dirname}/src/server`)],
          rollupConfig: {
            plugins: [
              typescriptPaths({
                tsConfigPath: 'tsconfig.base.json',
                preserveExtensions: true,
              }),
            ],
          },
...
```
`typescriptPaths` from `rollup-plugin-typescript-paths` is also required to make aliases work API handlers.

3. tRPC Routers
3.1. Each lib with tRPC support exposes their named routers in `modules/MY-MODULE/src/server/index.ts`
3.2. Each lib have their own tRPC client defined in `modules/MY-MODULE/src/trpc-client.ts`
3.3. tRPC routers are imported for each lib and merged in `apps/MY-APP/src/import-routes-trpc.ts`
3.4. Imported tRPC router is provided to the app in `apps/MY-APP/src/server/trpc-router.ts`. App specific routers are also merged here if there is any.

modules/notes/src/server/notes.ts:
```ts
import { z } from 'zod';
import { publicProcedure, router } from '@libs/trpc';
export const noteRouter = router({
...
});
```
modules/notes/src/server/index.ts:
```ts
import { router } from '@epmr/trpc';
import { noteRouter } from './notes';

export const mTRPCRouter = router({
  note: noteRouter,
});
```

apps/MY-APP/import-routes-trpc.ts
```ts
import { AnyRouter, combineRouters } from '@libs/trpc';

import { mTRPCRouter as mA_TRPCRouter } from '@modules/mod-a/server';
import { mTRPCRouter as mB_TRPCRouter } from '@modules/mod-b/server';
import { mTRPCRouter as mNotes_TRPCRouter } from '@modules/notes/server';

const TRPCRouters = [ mA_TRPCRouter, mB_TRPCRouter, mNotes_TRPCRouter ] as AnyRouter[];

export const TRPC_ROUTER = combineRouters( TRPCRouters );
```

libs/trpc/src/lib/combine-routers.ts:
```ts
import { AnyRouter, mergeRouters, router } from './trpc';

export function combineRouters(routers: AnyRouter[]) {
  return routers.reduce(
    (combined, router) => mergeRouters(combined, router),
    router({})
  );
}
```

apps/MY-APP/src/server/trpc-router.ts:
```ts
import { TRPC_ROUTER } from '../../__.import-routes-trpc';

export const appRouter = TRPC_ROUTER;

export type AppRouter = typeof appRouter;
```

What do you think?

**Comment by n-can-k** [comment]
If we can discover a more elegant solution for managing all 3 routers, I'm ready to take on the implementation and possibly develop a plugin for nx too

**Comment by brandonroberts** [comment]
> Thanks! I didn't know the ROUTES token functioned as a multi-token.
> 
> Looks like we just need to expose the `createRoutes` function for now.
> 
> Also, we're gonna need to expose a couple of types; `Files` (from './lib/routes') and `RouteExport` (from './lib/models'). They're needed to pass route objects between shared libraries and the app. Is it OK to go ahead with exposing these types?
> 
> With this setup, we can start using shared page routers that can be injected from the shared libs, which i think is pretty cool.

I think exporting `createRoutes`, `Files`, and `RouteExport` is a good start. We're moving the tRPC integration to  `@spartan-ng/trpc` maintained by @goetzrobin  but that should be a simple name change.

I've also looked at providing a way to configure multiple route page directories at the plugin level, but that's something for the future.

**Comment by n-can-k** [comment]
> I think exporting `createRoutes`, `Files`, and `RouteExport` is a good start.

Hey @brandonroberts 👋, just dropped a PR for this.
Let me know what you think!

**Comment by brandonroberts** [comment]
Included in the `0.2.44` release of Analog
